### ISSUE TITLE
partial_velocity should not require a rexpression in the provided reference frame

### ISSUE THREAD

**Issue by moorepants** [sympy/sympy#10865]
I think that we may have a fundamental design flaw with some things about physics.vector. To illustrate, let's say that we have defined a vector that describes the linear velocity of point P in the reference frame N in terms of the generalized speeds: u1, u2, and u3.

```python
N_v_P = u1 * A.x + u2 * A.y + u3 * A.z
```

Note that we have this velocity vector is expressed in a frame A which has an arbitrary relationship with the frame N.

If point P happens to be an essential point to form the equations of motion with Kane's method, then we will need to take the partial velocity of the vector with respect to the generalized speeds in an intertial reference frame. If N happens to be that inertial reference frame, then we should be able to do:

```python
N_v_P.diff(u1, N)
```
and this should return:

```python
A.x
```
The partial velocity of N_v_P in N with respect to u1 is simply A.x. Of course, if A is oriented with respect to N in some way, then A.x can be expressed in N's unit vectors, but it isn't necessary to do that to compute the partial velocity.

The following code shows that this does not work:

```
moorepants@garuda:pydy(error-messages)$ ipython
Python 3.5.1 |Continuum Analytics, Inc.| (default, Dec  7 2015, 11:16:01) 
Type "copyright", "credits" or "license" for more information.

IPython 4.1.2 -- An enhanced Interactive Python.
?         -> Introduction and overview of IPython's features.
%quickref -> Quick reference.
help      -> Python's own help system.
object?   -> Details about 'object', use 'object??' for extra details.

In [1]: import sympy.physics.mechanics as me

In [2]: u1, u2, u3 = me.dynamicsymbols('u1, u2, u3')

In [3]: N = me.ReferenceFrame('N')

In [4]: A = me.ReferenceFrame('A')

In [5]: N_v_P = u1 * A.x + u2 * A.y + u3 * A.z

In [7]: N_v_P.diff(u1, N)
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-7-7a11f17c841e> in <module>()
----> 1 N_v_P.diff(u1, N)

/home/moorepants/miniconda3/lib/python3.5/site-packages/sympy/physics/vector/vector.py in diff(self, wrt, otherframe)
    519                 outvec += Vector([(v[0].diff(wrt), otherframe)])
    520             else:
--> 521                 if otherframe.dcm(v[1]).diff(wrt) == zeros(3, 3):
    522                     d = v[0].diff(wrt)
    523                     outvec += Vector([(d, v[1])])

/home/moorepants/miniconda3/lib/python3.5/site-packages/sympy/physics/vector/frame.py in dcm(self, otherframe)
    405         if otherframe in self._dcm_cache:
    406             return self._dcm_cache[otherframe]
--> 407         flist = self._dict_list(otherframe, 0)
    408         outdcm = eye(3)
    409         for i in range(len(flist) - 1):

/home/moorepants/miniconda3/lib/python3.5/site-packages/sympy/physics/vector/frame.py in _dict_list(self, other, num)
    250             return outlist[0]
    251         raise ValueError('No Connecting Path found between ' + self.name +
--> 252                          ' and ' + other.name)
    253 
    254     def _w_diff_dcm(self, otherframe):

ValueError: No Connecting Path found between N and A

In [9]: me.partial_velocity([N_v_P], [u1, u2, u3], N)
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-9-56e053551919> in <module>()
----> 1 me.partial_velocity([N_v_P], [u1, u2, u3], N)

/home/moorepants/miniconda3/lib/python3.5/site-packages/sympy/physics/vector/functions.py in partial_velocity(vel_list, u_list, frame)
    564         pvlist = []
    565         for j in u_list:
--> 566             vel = i.diff(j, frame)
    567             pvlist += [vel]
    568         list_of_pvlists += [pvlist]

/home/moorepants/miniconda3/lib/python3.5/site-packages/sympy/physics/vector/vector.py in diff(self, wrt, otherframe)
    519                 outvec += Vector([(v[0].diff(wrt), otherframe)])
    520             else:
--> 521                 if otherframe.dcm(v[1]).diff(wrt) == zeros(3, 3):
    522                     d = v[0].diff(wrt)
    523                     outvec += Vector([(d, v[1])])

/home/moorepants/miniconda3/lib/python3.5/site-packages/sympy/physics/vector/frame.py in dcm(self, otherframe)
    405         if otherframe in self._dcm_cache:
    406             return self._dcm_cache[otherframe]
--> 407         flist = self._dict_list(otherframe, 0)
    408         outdcm = eye(3)
    409         for i in range(len(flist) - 1):

/home/moorepants/miniconda3/lib/python3.5/site-packages/sympy/physics/vector/frame.py in _dict_list(self, other, num)
    250             return outlist[0]
    251         raise ValueError('No Connecting Path found between ' + self.name +
--> 252                          ' and ' + other.name)
    253 
    254     def _w_diff_dcm(self, otherframe):

ValueError: No Connecting Path found between N and A
```

The problem is that the Vector object has no awareness of that it is a velocity of some point in some reference frame. Because of this, it doesn't really make sense to ask it to differentiate itself in the reference frame N.

But our vector differentiation in frames does work. It works because we *require* that there be a per-specified orientation between reference frames A and N. If this orientation is specified the correct answer is returned:

```
In [10]: q1, q2, q3 = me.dynamicsymbols('q1, q2, q3')

In [11]: A.orient(N, 'Body', (q1, q2, q3), 'XYZ')

In [12]: N_v_P.diff(u1, N)
Out[12]: A.x
```

This requires re-expression of the vector in the N frame so that the measure numbers in the N frame of N_v_P can be differentiated in that frame. But we know that re-expression is not necessary to get the correct result.

This comes up when you have ignorable coordinates. For example, if you want to study a satellite that is moving in inertial reference frame N, you can often ignore the 6 coordinates associated with the primary degrees of freedom up front. Currently, our software does not allow us to easily do this because of the above mentioned limitation.

Some thoughts:

- Maybe we need to define differentiation of the position/orientatoins, velocity, and accelerations at only on the Point and RefernceFrame objects because these have the full information about a given velocity vector, i.e. those objects know what frame the vector is with respect to. The vector by itself is not aware of what frame it is wrt.
- Maybe we need a "null" orientation option that lets you connect frames A and N but not specify the generalized coordinates and direction cosine matrix that relates the two. Then the vector.diff() function can deal with this properly.
- The fact that Vectors can differentiate themselves without know what frame they are defined wrt, seems mathematically incorrect.
- It is important to note that re-expression of a vector in another frame is different that knowing what frame a velocity vector is associated with.
- Do we need a VelocityVector object that knows more info like the frame the velocity is wrt or the associated points?
- I think Autolev properly tracks this with their syntax, for example in `N_v_P> = u1 * A1> + u2 * A2> + u3 * A3>` both the left and right hand sides have semantic meaning. This defines a velocity vector in N of P with that vector expression. All of that info must be stored in a data structure behind the scenes. They must have different types of vectors.

@sympy/mechanics What do you all think about this? I feel like we have an odd hole here that may take more than a hack to properly fix.

tl;dr Differentiation of vectors wrt to generalized speeds should not require that orientations be set between the frames that are associated with the unit vectors in the velocity vector expression and the frame that the velocity is defined it.

**Comment by moorepants** [comment]
Here are some thoughts on using `Point` to take partial velocities.

```python
>>> N = ReferenceFrame('N')
>>> A = ReferenceFrame('A')
>>> p = Point('P')
>>> p.set_vel(N, u1 * A.x + u2 * A.y)
>>> p.partial_velocity(u1, N)
A.x
>>> p.partial_velocity((u1, u2), N)
(A.x, A.y)
>>> p.partial_velocity(A, u1)
Error: Point P's velocity is not defined in the A frame so the partial cannot be taken.
# But what if you know the orientation of A in N?
>>> A.orient(N, 'Axis', (theta, N.x))
# The following should still fail because there is no way to know what the velocity of P in A is until it is defined. P in A could be zero, or it could be moving in A. Just because you can express the velocity of P in N in the A reference frame does mean you know the velocity of P in A. Right?
>>> p.partial_velocity(A, u1)
Error: Point P's velocity is not defined in the A frame so the partial cannot be taken without any orientation information.
>>> p.set_vel(A, u1**2 * A.x)
>>> p.partial_velocity(A, u1)
2 * u1 * A.x
```

A Point has the full set of information to say something about partial linear velocities (as does a ReferenceFrame and angular partial velocities).
```

**Comment by srjoglekar246** [comment]
I like the idea of having the functionality in Point. This way, the original method does not get modified. Would having a <code>try...except</code> block that simply differentiates the vector in the current frame (if not, then tries to re-express) do?

**Comment by moorepants** [comment]
But the other issue is that the current `partial_velocity` function is incorrect. For example, how can you take the partial velocity of a vector that isn't a velocity vector. That is currently possible. Also, our Vector.diff() requires that rotation matrices be defined for all frames in the chain, which is too strict when differentiating for partial velocities, for example.

I think we have a bigger issue to solve here. I don't quite have my head around it all yet.

Some thoughts:

Given this defined relationship:

```python
N = ReferenceFrame('N')
A = ReferenceFrame('A')
B = ReferenceFrame('B')

B.orient(A, 'Axis', (theta, A.x))

v = a * B.x + b * A.y + c * N.x + theta * N.y
```

If you try to differentiate this vector with respect to a, b, c or theta, in frames N, A, or B, the current code tries to ensure that the vector can be expressed in the desired frame so that any appearances of the variable in question will be explicitly shown in that frame. That makes sense and I suppose that is the only way that this can be computed without unevaluated derivatives. But it the vector is a velocity vector and we know what frame that velocity is defined with respect to, there is extra information that can be taken into account to avoid having to do this explicit re-expression.

With only the info provided above this is the correct answer:

```python
>>> v.diff(a, B)
B.x + c * Derivative(N.x, a, B) + theta * Derivative(N.y, a, B)
```

Because there is no relation defined between B and N, we should get unevaluated derivatives. The current code raises an error.

But if we know that v is the velocity of point P in frame B and that the generalized speeds are a, b, and c, then we know that a, b, and c will never be present in the rotation matrices among N, A, and B so:

```python
>>> v.diff(a, B)
B.x
```

An option for a fix to Vector would be to inform `Vector.diff()` about the variables that will never be present in rotation matrices. For example:

```python
>>> v.diff(a, B, not_in_dcm=(a, b, c))
B.x
```

Or we could have a subclass of Vector: `VelocityVector` that could have some awareness of what is a coordinate and what isn't.

Whatever we do to fix Vector.diff(), I think we still need to have partial_velocities operate on Points and ReferenceFrame ( or `VelocityVector`) instead of plain vectors as it doesn't make any sense to take the partial velocity of a non-velocity vector.

**Comment by gilbertgede** [comment]
Well, it's been a little while since I thought about this, but...

Going back to Kane's book, he defines the partial derivative of a vector in a frame as:
<img width="581" alt="kane" src="https://cloud.githubusercontent.com/assets/331109/13936903/4a8a90c2-ef7e-11e5-8bb3-edcde1506678.png">  
or rather, the sum of the partial derivatives of the scalar components (expressed in the target differentiation frame) wrt the differentiator.

I seem to recall having a conversation with Luke 5 or so years ago where he had an example of why you _had_ to convert it to the frame you're taking the derivative wrt, or you'd miss some edge cases. Also, I don't see a reason why it should be disallowed to compute the partial derivative of any vector.

One of the things I was very strict about when working on this stuff though was that a vector is a vector - a velocity vector is just that, a vector which _we_ have associated with the motion of some point. So I'm definitely against introducing a `VelocityVector`.

I guess what you're pointing out is that because we (the dynamicist) have additional information for the case where you are taking the partial velocity wrt a generalized velocity (which cannot be involved in defining the orientations), there is the potential for the result to be computed with less required information.

Regarding restricting the use of `partial_velocity` to points/frames - that seems OK to me. I don't remember how much `partial_velocity` is used internally, but I imagine the only downside we'd be introducing is a harder requirement that more points be supplied, instead of just velocities. Re-implementing some (potentially simpler) differentiation code in that function could allow for the functionality you're describing, maybe?

Maybe @hazelnusse could chime in and see if he remembers that edge case example I mentioned?

**Comment by moorepants** [comment]
I don't think our "partial derivative" function, i.e. `Vector.diff()`, is necessarily incorrect, but I think our notion of "partial velocity" isn't quite robust. When computing a partial velocity you have some extra information at hand that may provide useful.

> One of the things I was very strict about when working on this stuff though was that a vector is a vector - a velocity vector is just that, a vector which we have associated with the motion of some point. So I'm definitely against introducing a VelocityVector.

But a velocity vector has two more important pieces of information that a normal vector doesn't have: (1) it is a velocity and (2) the frame with which that velocity is expressed. We incapsulate these concepts in our Point and ReferenceFrame object, i.e. Point.vel(N) access a velocity vector that is wrt to frame N. But our normal vector class is unaware of those two things that may provide useful. Lastly, neither of these are aware of what variables in a vector expression happen to be a generalized coordinate and generalized speed, which can also be very useful.

> I guess what you're pointing out is that because we (the dynamicist) have additional information for the case where you are taking the partial velocity wrt a generalized velocity (which cannot be involved in defining the orientations), there is the potential for the result to be computed with less required information.

Yes, this is exactly what we do by hand for many problems.

> Re-implementing some (potentially simpler) differentiation code in that function could allow for the functionality you're describing, maybe?

Yes, that might be a solution. I think we should do away with the partial_velocity function or make it accept Points and ReferenceFrames instead of arbitrary vectors.

Here is a simple example that exposes how our partial_velocity functionality fails to compute the dynamical equations of motion of the simplest rigid body, yet it is easy to do "by hand".

```python
import sympy as sm
import sympy.physics.mechanics as me

# Frame A is attached to a simple satellite that is moving in the Newtonian
# frame N.
N = me.ReferenceFrame('N')
A = me.ReferenceFrame('A')

# The satellite has a mass center at Ao
Ao = me.Point('Ao')

# The generalized speeds are defined such that the measure numbers of the body
# fixed angular rate of A in N and the linear velocity of the mass center of Ao
# in N are simply the six generalized speeds.
u = me.dynamicsymbols('u:6')
A.set_ang_vel(N, u[0] * A.x + u[1] * A.y + u[2] * A.z)
Ao.set_vel(N, u[3] * A.x + u[4] * A.y + u[5] * A.z)

# Now we have a single symmetric rigid body for the satellite with a mass.
mA, Ix, Iy, Iz = sm.symbols('mA, Ix, Iy, Iz')
S = me.RigidBody('S', Ao, A, mA, (me.inertia(A, Ix, Iy, Iz), Ao))

# Now form the partial velocities. We know that these are simply:
p_par = [me.Vector(0), me.Vector(0), me.Vector(0), A.x, A.y, A.z]
p_frame = [A.x, A.y, A.z, me.Vector(0), me.Vector(0), me.Vector(0)]

# But our vector diff function fails because there is no orientation defined
# between A and N, even though we know that the generalized speeds will not be
# present in the rotation matrix that relates the two. This means that
# partial_velocity() and KanesMethod() will fail.
#p_par = [S.masscenter.vel(N).diff(u_i, N) for u_i in u]
#p_frame = [S.frame.ang_vel(N).diff(u_i, N) for u_i in u]

# Now form T* and R* for the single rigid body.
T_star = (-S.frame.ang_acc_in(N).dot(S.central_inertia) -
          S.frame.ang_vel_in(N).cross(S.central_inertia.dot(S.frame.ang_vel_in(N))))
R_star = -mA * S.masscenter.acc(N)

# Finally, form frstar to find the equations of motion of the system.
frstar = []
for p_f, p_p in zip(p_frame, p_par):
    frstar.append(p_f.dot(T_star) + p_p.dot(R_star))

# The accelerations are easily solved for:
sm.solve(frstar, [u_i.diff() for u_i in u])
```

**Comment by moorepants** [comment]
@monthubbard, Might you like to comment on this? This is the partial velocity issue that I discussed with you the other day, but this conversation is code heavy, so you might not be interested, fyi.

**Comment by moorepants** [comment]
I've proposed a fix at #10895 that is the simplest fix I can think of. I may open another pull request in the future to move the partial velocity functionality to Point and ReferenceFrame instead of having it operate on arbitrary vectors.
